### ISSUE TITLE
Enhancement: dependent types

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#773]
https://getpsalm.org/r/8753a0d501
```php
<?php
  function takesInt(int $p): void {}
  function takesBool(bool $p): void {}
  function takesString(string $p): void {}

  $coll = ["a" => 1, "b" => true, "c" => "z"];

  foreach ($coll as $k => $v) {
    if ("a" === $k) takesInt($v);
    if ("b" === $k) takesBool($v);
    if ("c" === $k) takesString($v);
  }
```
Here psalm knows that `$coll` is `array{a:int(1),b:true,c:string('z')}` and thus `$k` is `string('a'|'b'|'c')` and `$v` is `int|bool|string`. However it could probably also be deduced that `$v` is `bool` *when* `$k` is `'b'`, `string` when `$k` is `'c'` and so on.

**Comment by muglug** [comment]
I don't think this is really necessary. That's basically just bad code - it should be rewritten:

```php
 $coll = ["a" => 1, "b" => true, "c" => "z"];

  takesInt($coll["a"]);
  takesBool($coll["b"]);
  takesString($coll["c"]);
```

**Comment by weirdan** [comment]
`$coll` is not necessarily something that can be accessed using offset. Consider generators, for example. Those can return same key multiple times during iteration, so using a loop to process those totally make sense.

**Comment by muglug** [comment]
Yeah, but still not really a worthwhile road to go down, I think
